### PR TITLE
Enable client bootstrapping to receive OSCORE security information

### PR DESCRIPTION
**Pull request by rikard-sics**
This commit enables the client to receive OSCORE security context information from the bootstrap server during the bootstrapping process. The bootstrap server web UI was updated to accept settings for OSCORE under the LWM2M Server tab. Finally some changes were done in the client side code to support this functionality.

### CODE REVIEW

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java [hunk 1]
```diff
@@ -102,9 +102,13 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
 
                         boolean useOscore = oscoreObjLink.getObjectId() == OSCORE;
                         if (useOscore) {
+                            LOG.trace("Bootstrap connection is using OSCORE.");
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:105]: Maybe something like  : 
```suggestion
                            LOG.trace("Server {} is using OSCORE.", info);
```
if we add this trace should we also add the same kind of trace for PSK, RPK, X509.
(or we just remove this trace)

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java [hunk 2]
```diff
@@ -102,9 +102,13 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
...
+
                             // get corresponding oscore object
                             LwM2mObjectInstance oscoreInstance = oscores.getInstance(oscoreObjectInstanceId);
-                            LOG.trace("Bootstrap connection is using OSCORE.");
+                            if (oscoreInstance == null) {
+                                LOG.error("Failed to retrieve OSCORE object linked from BS security object");
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:110]: In this case line below will probably raise `NullPointerException` ?

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java [hunk 3]
```diff
@@ -145,9 +149,13 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
 
                     boolean useOscore = oscoreObjLink.getObjectId() == OSCORE;
                     if (useOscore) {
+                        LOG.trace("Registration connection is using OSCORE.");
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:152]: same remarks as above.

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java [hunk 4]
```diff
@@ -145,9 +149,13 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
...
+
                         // get corresponding oscore object
                         LwM2mObjectInstance oscoreInstance = oscores.getInstance(oscoreObjectInstanceId);
-                        LOG.trace("Registration connection is using OSCORE.");
+                        if (oscoreInstance == null) {
+                            LOG.error("Failed to retrieve OSCORE object linked from DM security object");
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:157]: same remarks as above.

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java [hunk 5]
```diff
@@ -102,9 +102,13 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
 
                         boolean useOscore = oscoreObjLink.getObjectId() == OSCORE;
```
Comment (sbernard31) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:103]: oscoreObjLInk can probably be null as OSCORE SECURITY MODE is optional.

So Regarding issue above maybe we can change the code like this ?  
```java 
// find associated oscore instance (if any)
LwM2mObjectInstance oscoreInstance = null;
ObjectLink oscoreObjLink = (ObjectLink) security.getResource(SEC_OSCORE_SECURITY_MODE)
        .getValue();
if (oscoreObjLink != null) {
    int oscoreObjectInstanceId = oscoreObjLink.getObjectInstanceId();
    if (!oscoreObjLink.isNullLink() && oscoreObjLink.getObjectId() != OSCORE) {
        LOG.warn(
                "The security object's 'OSCORE Security Mode' links to an incorrect object type.");
    } else {
        oscoreInstance = oscores.getInstance(oscoreObjectInstanceId);
        if (oscoreInstance == null) {
            // maybe we should use same log level for this one and the other just above
            LOG.error("Failed to retrieve OSCORE object linked from BS security object");
        }
    }
}

if (oscoreInstance != null) {
    info.useOscore = true;
    info.masterSecret = getMasterSecret(oscoreInstance);
    info.senderId = getSenderId(oscoreInstance);
    info.recipientId = getRecipientId(oscoreInstance);
    info.aeadAlgorithm = getAeadAlgorithm(oscoreInstance);
    info.hkdfAlgorithm = getHkdfAlgorithm(oscoreInstance);
    info.masterSalt = getMasterSalt(oscoreInstance);
```

(and same for register)

**Review by sbernard31 — COMMENTED** [review]
All the org.eclipse.server.bootstrap java changes will be obsolete with #883.

So maybe we should wait I integrate it before to merge this ? :thinking:

**Comment by sbernard31** [comment]
> All the org.eclipse.server.bootstrap java changes will be obsolete with #883.

#883 is not integrated in master.

So both solution : 
- either we integrate this  PR in `oscore` and then we rebase `oscore` on `master` (meaning several code to adapt)
- either we first rebase `oscore` on `master`, then you adapt the PR.

I could help if needed, just let me know.

**Comment by rikard-sics** [comment]
> All the org.eclipse.server.bootstrap java changes will be obsolete with #883.

Yes, I see.

> #883 is not integrated in master.
> 
> So both solution :
> 
> * either we integrate this  PR in `oscore` and then we rebase `oscore` on `master` (meaning several code to adapt)
> * either we first rebase `oscore` on `master`, then you adapt the PR.
> 
> I could help if needed, just let me know.

Right, so #883 is now integrated in master. I suppose there may not be any benefit in integrating this PR into 'oscore' first, so probably the second option is better. Shall I rebase my code in this PR on master or will you rebase the 'oscore' branch on master? (And then I rebase my code here on the updated 'oscore' branch).

**Comment by sbernard31** [comment]
I will try to rebase `oscore` branch on master.

**Comment by sbernard31** [comment]
Done. There was some easy to fix conflicts, so I guess I should not add any regression.
You can now rebase your work. If you have issue/question with the new code (BootstrapConfigurationStore), do not hesitate to ask.

**Comment by rikard-sics** [comment]
> Done. There was some easy to fix conflicts, so I guess I should not add any regression.
> You can now rebase your work. If you have issue/question with the new code (BootstrapConfigurationStore), do not hesitate to ask.

Thanks! Yeah I will start having a look and may come back with questions.

**Comment by rikard-sics** [comment]
I am closing this pull request to open a new separate one instead as that simplifies for Git operations and how I have structured the follow up code.

**Pull request closed without merging**